File: graphql-api/src/main/java/io/prizy/publicapi/graphql/scalars/InstantScalar.java
package io.prizy.publicapi.graphql.scalars;

import graphql.language.StringValue;
import graphql.schema.Coercing;
import graphql.schema.CoercingParseLiteralException;
import graphql.schema.CoercingParseValueException;
import graphql.schema.CoercingSerializeException;
import graphql.schema.GraphQLScalarType;
import lombok.experimental.UtilityClass;

import java.time.Instant;

/**
 * @author Nidhal Dogga
 * @created 6/24/2022 9:19 PM
 */


@UtilityClass
public class InstantScalar {

  public GraphQLScalarType SCALAR = GraphQLScalarType.newScalar()
    .name("DateTime")
    .description("Timezoned date and time")
    .coercing(new InstantCoercing())
    .build();

  private class InstantCoercing implements Coercing<Instant, String> {

    @Override
    public String serialize(Object dataFetcherResult) throws CoercingSerializeException {
      if (dataFetcherResult instanceof Instant instant) {
        return instant.toString();
      }
      throw new CoercingSerializeException("Cannot serialize: " + dataFetcherResult);
    }

    @Override
    public Instant parseValue(Object input) throws CoercingParseValueException {
      return Instant.parse(input.toString());
    }

    @Override
    public Instant parseLiteral(Object input) throws CoercingParseLiteralException {
      if (input instanceof StringValue stringValue) {
        return Instant.parse(stringValue.getValue());
      }
      throw new CoercingParseLiteralException("Cannot parse value: " + input);
    }
  }

}

File: public-api/src/main/kotlin/io/prizy/publicapi/graphql/scalars/InstantScalar.kt
package io.prizy.publicapi.graphql.scalars

import graphql.language.StringValue
import graphql.schema.Coercing
import graphql.schema.CoercingParseLiteralException
import graphql.schema.CoercingSerializeException
import graphql.schema.GraphQLScalarType
import java.time.Instant

object InstantScalar {

  val graphqlScalarType: GraphQLScalarType = GraphQLScalarType.newScalar()
    .name("Instant")
    .description("Timezoned date and time")
    .coercing(InstantCoercing)
    .build()

  object InstantCoercing : Coercing<Instant, String> {

    override fun serialize(dataFetcherResult: Any): String {
      when (dataFetcherResult) {
        is Instant -> return dataFetcherResult.toString()
        else -> throw CoercingSerializeException("Cannot serialize $dataFetcherResult")
      }
    }

    override fun parseValue(input: Any): Instant {
      return Instant.parse(input.toString())
    }

    override fun parseLiteral(input: Any): Instant {
      when (input) {
        is StringValue -> return Instant.parse(input.value)
        else -> throw CoercingParseLiteralException("Cannot parse value: $input")
      }
    }
  }
}